go/cmd/recordchecker/recordchecker.go

```diff
@@ -0,0 +1,313 @@
+	_, err := cl.RunInTransaction(ctx, func(tx *datastore.Transaction) error {
+		lastRunKey := datastore.NameKey(jobDataKind, JobDataLastRun, nil)
+		lr := jobDataLastRunEntity{Value: data.lastRun}
+		_, err := tx.Put(lastRunKey, &lr)
+		if err != nil {
+			return fmt.Errorf("failed to write %s: %w", JobDataLastRun, err)
+		}
+
+		invalidRecordsKey := datastore.NameKey(jobDataKind, JobDataInvalidRecords, nil)
+		ir := jobDataInvalidRecordsEntity{Value: data.invalidRecords}
+		_, err = tx.Put(invalidRecordsKey, &ir)
+		if err != nil {
+			return fmt.Errorf("failed to write %s: %w", JobDataInvalidRecords, err)
+		}
+
+		return nil
+	})
+
+	return err
+}
+
+type checkTask struct {
+	ID   string
+	Vuln *models.Vulnerability
+}
+
+// run is the main application logic
+func run(ctx context.Context, env *appEnv) error {
+	rcData, err := getRecordCheckerData(ctx, env.ds)
+	if err != nil {
+		return fmt.Errorf("failed to get prior run data: %w", err)
+	}
+
+	tasksChan := make(chan checkTask)
+	resultsChan := make(chan checkRecordResult)
+
+	// Start the results handler
+	var resultsWg sync.WaitGroup
+	resultsWg.Add(1)
+	var newInvalid []string
+	go func() {
+		defer resultsWg.Done()
+		for result := range resultsChan {
+			if handleResult(ctx, env.topic, result) {
+				newInvalid = append(newInvalid, result.ID)
+			}
+		}
+	}()
+
+	// Start the worker pool
+	var workerWg sync.WaitGroup
+	for range env.numWorkers {
+		workerWg.Add(1)
+		go func() {
+			defer workerWg.Done()
+			for task := range tasksChan {
+				checkRecord(ctx, env.ds, env.bucket, task.ID, task.Vuln, resultsChan)
+			}
+		}()
+	}
+
+	// Queue all invalid records from the previous run.
+	for _, id := range rcData.invalidRecords {
+		logger.Debug("checking previously invalid record", slog.String("id", id))
+		tasksChan <- checkTask{ID: id}
+	}
+
+	// Queue all new records.
+	runStartTime := time.Now().UTC()
+	query := datastore.NewQuery("Vulnerability")
+	if rcData.lastRun != nil {
+		query = query.FilterField("modified", ">", *rcData.lastRun)
+	}
+	it := env.ds.Run(ctx, query)
+	for {
+		var vuln models.Vulnerability
+		key, err := it.Next(&vuln)
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("failed to query vulnerabilities: %w", err)
+		}
+		logger.Debug("checking newly modified record", slog.String("id", key.Name))
+		tasksChan <- checkTask{ID: key.Name, Vuln: &vuln}
+	}
+	// Wait for all tasks to finish processing
+	close(tasksChan)
+	workerWg.Wait()
+	close(resultsChan)
+	resultsWg.Wait()
+
+	// Update the record checker run data
+	rcData = recordCheckerData{
+		lastRun:        &runStartTime,
+		invalidRecords: newInvalid,
+	}
+	if err := writeRecordCheckData(ctx, env.ds, rcData); err != nil {
+		return fmt.Errorf("failed to store run data: %w", err)
+	}
+
+	return nil
+}
+
+func handleResult(ctx context.Context, topic *pubsub.Topic, result checkRecordResult) bool {
+	wasInvalid := false
```

Comment:
Can you add a comment here (or just on the function) about what the return type of wasInvalid is suppose to mean? It's a bit unclear why NeedsRetry makes it invalid.